### PR TITLE
Update the UI of the Juju Web CLI

### PR DESCRIPTION
**Pull request by hatched**
## Done

- Update the UI of the Juju Web CLI to the latest design.
- Removed the old `Terminal` component and dependencies.

## QA

- Bootstrap a new controller using the Juju 2.9 snap
  - `snap refresh --channel latest/beta`
- Build this PR and follow [this guide to connect to your controller](https://github.com/canonical-web-and-design/jaas-dashboard#developing-while-connected-to-a-juju-controller)
- View your controller model and the terminal should appear along the bottom.
- Steps:
  - - [ ] Click the help icon, did it show the help?
  - - [ ] Type `status` in the prompt and submit, did the model status appear?
  - - [ ] Type `status --color` in the prompt and submit, did the old status get cleared and a new coloured one appear?
  - - [ ] Type `help commands` in the prompt and submit, can you scroll the output
  - - [ ] Click and drag along the top of the output, does it resize as expected
  - - [ ] Drag it to the bottom so it's "closed"
  - - [ ] Type `status` in the prompt and submit, did it re-open?
  - - [ ] Click the help icon, did it resize and show the help content?
  - - [ ] Chaos monkey the interactions, can you find any bugs? Improvements?
  - - [ ] Chaos monkey the juju commands, can you find any bugs? Improvements?

[Zeplin Designs](https://app.zeplin.io/project/5f6234f4cb1e87aac4650d25/dashboard?sid=5f6235582ba7c6a777814c7f)

## Details

Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1445
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1446
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1447
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1448
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1456
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1396

## Screenshots
Screen Shot 2020-10-21 at 12 07 20 PM

### CODE REVIEW

File: src/components/WebCLI/Output.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  0: null, // reset
+  30: "0,0,0", // Black
+  31: "205,49,49", // Red
+  32: "13,188,121", // Green
+  33: "229,229,16", // Yellow
+  34: "26,114,200", //  Blue
+  35: "188,63,188", // Magenta
+  36: "17,168,205", // Cyan
+  37: "229,229,229", // White
+  90: "102,102,102", // Bright Black (Gray)
```
Comment (anthonydillon) [inline src/components/WebCLI/Output.js:15]: Haha, Bright Black (Gray)

File: src/components/WebCLI/_webcli.scss [hunk 2]
```diff
@@ -0,0 +1,90 @@
+@import "../../scss/settings";
+@import "vanilla-framework/scss/vanilla";
+
+.webcli {
+  bottom: 0;
+  position: fixed;
+  width: calc(100vw - #{$sidebar-collapsed});
+
+  &__input {
+    background-color: black;
```
Comment (anthonydillon) [inline src/components/WebCLI/_webcli.scss:10]: `$color-x-dark;`

File: src/components/WebCLI/_webcli.scss [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      line-height: 2.5rem;
+      padding-right: 0.5rem;
+    }
+
+    form {
+      flex-grow: 1;
+    }
+
+    &-input[type="text"] {
+      background-color: black;
```
Comment (anthonydillon) [inline src/components/WebCLI/_webcli.scss:32]: `$color-x-dark`

File: src/components/WebCLI/_webcli.scss [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      color: $color-mid-light;
+      // https://github.com/canonical-web-and-design/vanilla-framework/pull/3370
+      // XXX The unquote will need to be removed with the next vanilla release
+      font-family: unquote($font-monospace);
+      height: 2.5rem;
```
Comment (anthonydillon) [inline src/components/WebCLI/_webcli.scss:37]: Maybe set the height (2.5rem) to a variable so you can update the height from a single variable at the top of the file.

File: src/components/WebCLI/_webcli.scss [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  width: calc(100vw - #{$sidebar-collapsed});
+
+  &__input {
+    background-color: black;
+    display: flex;
+    height: 2.5rem;
+    padding: 0 1rem;
+
+    &-prompt {
+      color: #39b54a;
```
Comment (anthonydillon) [inline src/components/WebCLI/_webcli.scss:16]: Can you make this color a variable at the top of the document

File: src/components/WebCLI/_webcli.scss [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      margin-left: 1rem;
+
+      i {
+        cursor: pointer;
+      }
+    }
+  }
+
+  &__output {
+    background-color: #2c001e;
```
Comment (anthonydillon) [inline src/components/WebCLI/_webcli.scss:54]: Would be good to have a nicely named variable for this as a setting at the top of this file

**Review by anthonydillon — COMMENTED** [review]
Looking awesome @hatch few suggestions in the code.

**Comment by spencerbygraves** [comment]
@hatched thanks, it's looking good. I noticed a couple of things, hope they make sense.

- If I click help and then drag it closed again, clicking the help icon doesn't open it again
- After entering `help commands` I scrolled to the bottom. When I then drag the window down to close it again, the content scrolls back up. Is that expected?
- If I open help and drag the window up a bit, when I then type 'status' it doesn't resize the window to fit it all in, some is lost at the bottom.
- Could the window only open up as much as it needs to? After I enter `status` there is quite a bit of space under the text
- There is some inconsistent text positioning, the help text is closer to the top of the window than when you input a command. Could you try to match the design please.

image

- Could this menu use the [contextual menu pattern](https://vanillaframework.io/docs/examples/patterns/contextual-menu/default). Also, that menu doesn't display the other commands I entered after status.

image

**Comment by hatched** [comment]
>     * If I click help and then drag it closed again, clicking the help icon doesn't open it again

This is a bug, will fix

>     * After entering `help commands` I scrolled to the bottom. When I then drag the window down to close it again, the content scrolls back up. Is that expected?

I'm not able to reproduce this, but I have seen it happen before, I'll see if I can find a repro case for the 'bounce to full screen' issue.

>     * If I open help and drag the window up a bit, when I then type 'status' it doesn't resize the window to fit it all in, some is lost at the bottom.

This was intentional as I treated the value you set it to to be more important than what we wanted to set it to.

>     * Could the window only open up as much as it needs to? After I enter `status` there is quite a bit of space under the text

I played around with this but wasn't able to create a pleasing and reliable experience within the timebox, I'll create an issue to see if we can do this properly in a follow-up.

>     * There is some inconsistent text positioning, the help text is closer to the top of the window than when you input a command. Could you try to match the design please.

So the spacing is actually the same but the responses from the backend sometimes have an extra newline - I've filed a bug report with them to resolve that. Good catch :)

>     * Could this menu use the [contextual menu pattern](https://vanillaframework.io/docs/examples/patterns/contextual-menu/default). Also, that menu doesn't display the other commands I entered after status.

As discussed offline this isn't part of the dashboard, maybe an autocomplete in your browser, I've disabled native autocomplete on the input to try and disable that.

File: src/components/WebCLI/Output.js (hunk 1 above)
Comment (hatched) [reply src/components/WebCLI/Output.js:15]: Haha yeah I copied the description from the wikipedia article above, I also thought it was a funny inclusion :D

File: src/components/WebCLI/_webcli.scss (hunk 4 above)
Comment (hatched) [reply src/components/WebCLI/_webcli.scss:37]: done

File: src/components/WebCLI/_webcli.scss (hunk 6 above)
Comment (hatched) [reply src/components/WebCLI/_webcli.scss:54]: done

**Comment by hatched** [comment]
@spencerbygraves @clagom @anthonydillon this is ready for some more eyes. I'm still working on the tests but I believe I've addressed all the issues (and then some) or created follow-ups for the others.

File: src/components/WebCLI/Output.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+const DEFAULT_HEIGHT = 300;
+
+const WebCLIOutput = ({ content, helpMessage, showHelp }) => {
+  const resizeDeltaY = useRef(0);
+  const [height, setHeight] = useState(1);
+
+  useEffect(() => {
+    const resize = (e) => {
+      const viewPortHeight = window.innerHeight;
+      const mousePosition = e.clientY + 40; // magic number
```
Comment (barrymcgee) [inline src/components/WebCLI/Output.js:78]: Why is it a magic number?

**Comment by spencerbygraves** [comment]
Thanks @hatched. My first two points above are fixed (open/close help and scrolling).

I still think number 3 is important. For example, If I click help and then straight after type `status`, I can't see the result without dragging the window up. I think this could be the first experience people have when using it.

That aside I'll give a design +1 as you're going to follow up on the spacing issue.

Cheers.

**Comment by barrymcgee** [comment]
The bottom of the terminal needs bottom padding to account for the command prompt - `whoami` is obscured

Screenshot 2020-10-23 at 09 45 35

**Review by barrymcgee — CHANGES_REQUESTED** [review]

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "juju-dashboard",
-  "version": "0.3.0",
```
Comment (barrymcgee) [inline package.json:3]: This doesn't seem reflected in the yarn.lock file?

**Comment by barrymcgee** [comment]
Clicking to the right of the terminal content can incrementally reduce the height of the terminal.

Screen Recording 2020-10-23 at 09 56 29 am

**Comment by barrymcgee** [comment]
Clicking whitespace under model details tables can maximise terminal unexpectedly

Screen Recording 2020-10-23 at 10 02 16 am

File: src/components/WebCLI/_webcli.scss [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    a {
+      text-decoration: underline;
+    }
+
+    code,
+    pre {
+      background-color: $canonical-purple;
+      border: none;
+      color: white;
```
Comment (barrymcgee) [inline src/components/WebCLI/_webcli.scss:69]:
```suggestion
      color: $color-x-light;
```

File: src/components/WebCLI/WebCLI.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+  /*
+    The messageBuffer is required because the websocket returns messages much
+    faster than React wants to update the component. Doing this allows us to
+    store the messages in a buffer and then set the output every cycle.
+  */
+  const messageBuffer = useCallback(
+    (message) => {
+      wsMessageStore.current = wsMessageStore.current + message;
+      setTimeout(() => {
```
Comment (barrymcgee) [inline src/components/WebCLI/WebCLI.js:40]: Doesn't this setTimeout require a timeout value?

**Comment by spencerbygraves** [comment]
One other thing, as you can drag the shell from any point, not just the drag bar, should the cursor be different?

File: src/components/WebCLI/WebCLI.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        content={output}
+        showHelp={shouldShowHelp}
+        helpMessage={`Welcome to the Juju Web CLI - see the <a href="https://juju.is/docs/webcli" class="p-link--inverted" target="_blank">full documentation here</a>.`}
+      />
+      <div className="webcli__input">
+        <div className="webcli__input-prompt">$ juju</div>
+        <form onSubmit={handleCommandSubmit}>
+          <input
+            autoComplete="off"
+            className="webcli__input-input"
```
Comment (barrymcgee) [inline src/components/WebCLI/WebCLI.js:132]: 🤨

File: src/components/WebCLI/Output.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    colorizedContent = colorizedContent + part;
+    previousIndex = color.index;
+  });
+  return colorizedContent;
+};
+
+const DEFAULT_HEIGHT = 300;
+
+const WebCLIOutput = ({ content, helpMessage, showHelp }) => {
+  const resizeDeltaY = useRef(0);
```
Comment (barrymcgee) [inline src/components/WebCLI/Output.js:72]: AFAIK a `useRef` only ever returns an object so this default should be `null`, not `0`.

File: src/pages/Models/Details/ModelDetails.js [hunk 13]
```diff
@@ -376,7 +383,13 @@ const ModelDetails = () => {
...
         </div>
       )}
-      {generateTerminalComponent(modelUUID, controllerWSHost)}
+      {showWebCLI ? (
+        <WebCLI
+          controllerWSHost={controllerWSHost}
+          credentials={credentials}
+          modelUUID={modelUUID}
+        />
+      ) : null}
```
Comment (barrymcgee) [inline src/pages/Models/Details/ModelDetails.js:392]: If you have nothing to show as an alternative, you can simplify this conditional like so;

```
      {showWebCLI &&
        <WebCLI
          controllerWSHost={controllerWSHost}
          credentials={credentials}
          modelUUID={modelUUID}
        />
     }
```

**Comment by anthonydillon** [comment]
QA issue when trying to copy a public address:
Peek 2020-10-23 10-54

**Comment by hatched** [comment]
Sorry folks I pushed the wrong tarball on my way out the door last night, The one up there now has all of these issues resolved.

File: src/components/WebCLI/Output.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  const [height, setHeight] = useState(1);
+
+  useEffect(() => {
+    const resize = (e) => {
+      const viewPortHeight = window.innerHeight;
+      const mousePosition = e.clientY + 40; // magic number
+      const newHeight = viewPortHeight - mousePosition + resizeDeltaY.current;
+
+      const maximumOutputHeight = viewPortHeight - 100; // magic number.
```
[the remark this answers is not in the document]
Comment (hatched) [reply src/components/WebCLI/Output.js:81]: It's just what felt right spacing wise

File: src/components/WebCLI/Output.js (hunk 7 above)
Comment (hatched) [reply src/components/WebCLI/Output.js:78]: It's what felt right when dragging things around.

File: src/components/WebCLI/WebCLI.js (hunk 10 above)
Comment (hatched) [reply src/components/WebCLI/WebCLI.js:40]: When you use setTimeout without a time it effectively pops it to the top of the stack to be executed on the next loop.

File: src/components/WebCLI/WebCLI.js (hunk 11 above)
Comment (hatched) [reply src/components/WebCLI/WebCLI.js:127]: it's an inputception

File: src/components/WebCLI/Output.js (hunk 12 above)
Comment (hatched) [reply src/components/WebCLI/Output.js:81]: It returns an object `{current: <initialValue>}` and the value passed to the `useRef()` sets that `initialValue`. So in this case I wanted the delta to start off as being 0.

File: src/pages/Models/Details/ModelDetails.js (hunk 13 above)
Comment (hatched) [reply src/pages/Models/Details/ModelDetails.js:392]: done

**Comment by hatched** [comment]
OK This one is now all ready to go, QA environment has been updated with the latest build.

**Comment by anthonydillon** [comment]
Thanks @hatched this is looking great. I am testing on my phone and here are a few things I found. None are blockers so feel free to fix in a follow up.

- Add `autocapitalize="off"` to the input__input. (Sorry, in joke). It's annoying for phones to capitalise commands by default
- Can you add an analytics event to command. I think that will help us make decisions on what users want.
- Can you add touch events to the resizing code as I can't resize the CLI on my phone

**Comment by hatched** [comment]
Thanks for the input @anthonydillon I've resolved the first two, but created an issue for the last one as it appears to be a bit bigger of an issue.

**Comment by ziheliu214** [comment]
@hatched Thanks Jeff amazing work!

UX questions:

1) Was the output designed to be one-off record, that every time jumping to different pages (including clicking on the help) it would be erased and not be able to scroll to historical output?

2)
image
When a wrong command is entered, it suggests to try `run` but I am not sure if it is supported.

**Comment by hatched** [comment]
Thanks @ziheliu214 
1. Yes the output intentionally clears after every new command submission. This was so that it was more clear that this isn't a 'shell' but a way to run juju cli commands. We can adjust this easily so lets wait to hear back from users what they think.
2. It looks like you submitted an empty command, We should probably prevent that. The output you're seeing is directly from Juju. It's trying to hint to you what command you maybe meant to run.

**Review by anthonydillon — APPROVED** [review]
Code and QA :+1: great work @hatched

**Review by barrymcgee — APPROVED** [review]

**Comment by hatched** [comment]
Thanks all for the reviews and qa's!

**Pull request merged**